Feature recipe modal

MinaKhamesi opened the pull request:
Designed and built modal that will allow a chef to add/edit/delete a recipe
Refactored controller for recipe endpoints (add/edit/delete) 
Added context for recipes so that when a recipe is added or edited or deleted, list of recipes on Chef profile page will get updated.
For adding and testing these functionalities, I needed some test components, so TestChefProfile component and Recipe component are both for this purpose and were not meant to be used in final product. 
for testing the functionalities, you need to login and set up a chef profile. I edited the App.js to automatically bring the user to the testChefComponent, in there you have a button that will allow the user to mark himself/herself as a chef and create a chefProfile. Since the profile context is not in place you need to refresh the page to see the component that will allow to see the modal by clicking on the add button to create a new recipe. One recipes added you can click on the edit button to see the modal in edit format, where the functionalities for editing the recipe and deleting the recipes are there.



<img width="1440" alt="Screen Shot 2021-03-18 at 9 02 01 AM" src="https://user-images.githubusercontent.com/53087960/111630876-30183080-87c9-11eb-9ad7-37b1bd8e9cd5.png">
<img width="1440" alt="Screen Shot 2021-03-18 at 9 02 44 AM" src="https://user-images.githubusercontent.com/53087960/111630911-37d7d500-87c9-11eb-8888-e70b946b5532.png">
<img width="1440" alt="Screen Shot 2021-03-18 at 9 03 06 AM" src="https://user-images.githubusercontent.com/53087960/111630945-3dcdb600-87c9-11eb-99f8-7faf58d9384a.png">
<img width="1440" alt="Screen Shot 2021-03-18 at 9 03 34 AM" src="https://user-images.githubusercontent.com/53087960/111630961-43c39700-87c9-11eb-898b-19acf8bd4606.png">
@dproc96

MinaKhamesi committed:
Add modals for edit profile and add/edit recipe

MinaKhamesi committed:
Add Recipe context functionality

MinaKhamesi committed:
Add a test Recipe component and modify the recipe modal to prefill the information for editting a recipe

MinaKhamesi committed:
Add id in the action for now, to be able handle edit action

MinaKhamesi committed:
Resolve Merge conflicts

MinaKhamesi committed:
Delete a test route that no longer is needed

MinaKhamesi committed:
Incorporate backend routes into context actions

MinaKhamesi committed:
Merge dev into the branch & resolve conflicts

MinaKhamesi committed:
Modify backend to have a functional modal with logical behaviour

Added isChef route and controller, modifield the create chefProfile route to save only userId in the user field, modified recipeMiddleWare so that it won't populate user field in every route because it was causing problem when we wanted to edit a recipe. it will save the whole user model in the user field, Also added timestamp to recipe model so that when user is fetching the recipes it will be sorted by created time, and getAllRecipesByChef was modified to fetch the correct data, before it was fetching all recipes and since recipes had wrong user field because of the malfunctional middleWare the route had weird behaviour, sometimes returning information of other chefs only.

MinaKhamesi committed:
Add logout route and modify the frontend accordingly

MinaKhamesi committed:
Change routes to have array of information

MinaKhamesi committed:
Refactore action and constant naming and undo the changes in chefProfile middleware and getRecipeByChefId route

MinaKhamesi committed:
Merge dev into branch

MinaKhamesi committed:
Merge dev into branch and resolve conflicts

MinaKhamesi committed:
Delete a test component that no longer is needed

WhoaNellie commented on client/src/actions/recipeActions.js:
```diff
@@ -0,0 +1,108 @@
+    };
+    const res = await axios.post("/api/recipes", formData, config);
+    if (res.status === 201) {
+      console.log(res.data.newRecipe);
```
Console logs like this can be removed

WhoaNellie commented on client/src/components/Dialogs/RecipeModal.js:
```diff
@@ -0,0 +1,178 @@
+    fontSize: 12,
+  },
+  input: {
+    margin: "3px 0",
```
Try to use theme.spacing if you can, it can take decimal values

WhoaNellie commented on client/src/components/Formik/Input.js:
```diff
@@ -39,7 +39,13 @@ const Input = ({ label, name, ...rest }) => {
               {...field}
-              helperText={form.touched[name] ? form.errors[name] : ""}
+              helperText={
+                form.touched[name]
+                  ? form.errors[name]
+                  : helperText
+                  ? helperText
+                  : ""
```
Pulling out some nested ternary statements into variables could make sections like this more readable

WhoaNellie commented on client/src/context/recipe-context.js:
```diff
@@ -0,0 +1,28 @@
+const RecipeContextProvider = ({ children }) => {
+  const [state, dispatch] = useReducer(RecipeReducer, initialState);
+
+  // const addRecipe = (recipe) => {
+  //   dispatch({ type: "AddRecipe", payload: recipe });
+  //   // setRecipes([...recipes, recipe]);
+  // };
+
+  // const removeRecipe = (id) => {
+  //   // setRecipes(recipes.filter((recipe) => recipe.id !== id));
+  // };
```
Can remove this and other commented out code

WhoaNellie commented on client/src/pages/TestChefProfile.js:
```diff
@@ -0,0 +1,149 @@
+          variant="contained"
+          color="secondary"
+          onClick={handleNewChef}
+          style={{
+            borderRadius: "0",
+            width: "200px",
+            height: "40px",
+            marginTop: "30px",
+            textTransform: "capitalize",
```
Try to avoid inline styling

MinaKhamesi replied:
Hey Nellie, Thanks for your time. I am working on the comments. 
Just wanted to say, this is a test component. Because at the time I was working on recipe functionality, there was no chefProfile, I had to have a basic one to see my changes and my functionality in it-There still is no chef profile in dev-. So basically only the logic part of the component will remain(The logic for context and actions and so on. So if that's okay, can i ignore the parts that is not going to be in the final application. I will completely delete these test components when Val made his merge and I added the functionalities I want to his component.

WhoaNellie reviewed (commented):
This is a big PR but don't forget to clean up things like commented out code and console logs 👍

jessdesr commented on client/src/actions/userActions.js:
```diff
@@ -101,8 +101,8 @@ export const register = async (dispatch, registerPayload) => {
   dispatch({
     type: USER_LOGOUT,
   });
+  await axios.get("/api/users/logout");
```
You'll probably want to not remove the userInfo from localstorage if this request fails

jessdesr commented on client/src/components/Recipe.js:
```diff
@@ -0,0 +1,163 @@
+import { makeStyles } from "@material-ui/core/styles";
+import clsx from "clsx";
+import Card from "@material-ui/core/Card";
+import CardHeader from "@material-ui/core/CardHeader";
```
You can import these material ui components using destructuring, so instead of having to have import statements per component, you can for example:
`import { Card, CardHeader, CardMedia, etc, etc } from "@material-ui/core`

MinaKhamesi replied:
Hey @jessdesr Thanks for your time. I'm working on the comments. Just wanted to ask since this component is a test component and will be replaced with Val's component for recipe, is it okay to ignore the none logic related parts? I mean, I actually just copy past a Card component from material ui and inserted the information of recipes in the elements to show that the context and the actions are working. I didn't give the rest of the parts that much thought.

MinaKhamesi replied:
But generally, for my own components I try to write the code as you said.

jessdesr commented on client/src/components/Recipe.js:
```diff
@@ -0,0 +1,163 @@
+      if (srcData || !pictureKey) return;
+      const response = await getSrcData(pictureKey);
+      if (response.srcData) {
+        // setSrc(response.srcData);
```
clear out this commented code

jessdesr commented on server/controllers/userController.js:
```diff
@@ -81,5 +81,31 @@ const loginUser = AsyncHandler(async (req, res) => {
+const logoutUser = async (req, res) => {
+  // Set token to none and expire after 5 seconds
+  console.log("we are in controller");
+  res.cookie("token", "none", {
```
You should delete the cookie instead of creating another token that expires

jessdesr commented on server/routes/userRoutes.js:
```diff
@@ -1,9 +1,17 @@
 router.route("/register").post(registerUser);
 router.route("/login").post(loginUser);
+router.route("/logout").get(logoutUser);
+router.route("/isChef").put(auth, makeUserAChef);
```
maybe rename this, I would assume this route would return whether or not a user is a chef, not make the user a chef

jessdesr reviewed (approved):
Overall nice work, cover Nellie's comments as well as my suggestions then it can be merged to dev!  

I noticed you've also merged into the demo branch: make sure once this is merged to dev you also update the demo branch.

MinaKhamesi committed:
Merge branch 'dev' into feature-recipe-modal
Merge Dev to Branch

MinaKhamesi committed:
Clean the files, get rid off comments and console logs

MinaKhamesi committed:
Delete unnecessary lines of code for styling

MinaKhamesi committed:
Fix logout route and action

Logout route now will clear the cookie, before it only expired it, in the action it will wait for the response from backend and only if successfully logged out the userInfo will get clear, also the route '/api/users/isChef for marking a user as a chef has a new naming '/api/users/markChef'

MinaKhamesi committed:
Add frontend validation for picture is required

MinaKhamesi committed:
Add validation to price to accept only up to two decimals

MinaKhamesi committed:
Change Recipe components name to TestRecipe

MinaKhamesi committed:
Add RECIPE_RESET action and reducer

when user logged out and the next user logged in and the new user had no recipe, the recipe of previous user was showing so I added this new action to fire up every time a user loges out so that we have an empty recipe array by default in this situation
